File: 92. Reverse Linked List II/main.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func genFromSlice(s []int) *ListNode {
	res := ListNode{}
	nexPtr := &res

	for _, val := range s {
		nexPtr.Next = &ListNode{
			Val: val,
		}
		nexPtr = nexPtr.Next
	}

	return res.Next
}

func printListNode(node *ListNode) {
	curr := node
	for curr != nil {
		fmt.Println(curr.Val)
		curr = curr.Next
	}
}

func main() {
	node := genFromSlice([]int{3})
	ans := reverseBetween(node, 1, 1)
	printListNode(ans)
}

// ver1
/* func reverseBetween(head *ListNode, left int, right int) *ListNode {
	curr := head
	lBoundary := head
	for i := 1; i < left; i++ {
		lBoundary = curr
		curr = curr.Next
	}

	newHead, _ := reverse(curr, right-left)

	if left == 1 {
		head = newHead
	} else {
		lBoundary.Next = newHead
	}

	return head
} */

// ver 2
func reverseBetween(head *ListNode, left int, right int) *ListNode {
	newHead := &ListNode{
		Next: head,
	}

	curr := newHead
	for i := 0; i < left-1; i++ {
		curr = curr.Next
	}

	lHead, rBoundary := reverse(curr.Next, right-left)
	curr.Next.Next = rBoundary
	curr.Next = lHead

	return newHead.Next
}

func reverse(head *ListNode, count int) (*ListNode, *ListNode) {
	if head.Next == nil || count == 0 {
		return head, head.Next
	}
	newParent, rBoundary := reverse(head.Next, count-1)
	head.Next.Next = head
	return newParent, rBoundary
}

File: 92. Reverse Linked List II/main.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        i = 1
        ptr = ListNode(0, head)
        while i < left:
            ptr = ptr.next
            i+=1

        reversePtr = ptr.next
        last = ptr
        nextPtr = ptr
        while i <= right:
            nextPtr = reversePtr.next
            reversePtr.next = last
            last = reversePtr
            reversePtr = nextPtr
            i+=1
        
        ptr.next.next = nextPtr
        ptr.next = last

        return head

s = Solution()
l = ListNode(1,ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
s.reverseBetween(l,2,4)
print("")